### ISSUE TITLE
TaxonWorks download importing incorrectly into ChecklistBank

### ISSUE THREAD

**Issue by dhobern** [CatalogueOfLife/backend#1242]
The latest version of the TaxonWorks Global Lepidoptera Index dataset gets fragmented into many siblings of Lepidoptera on import into CLB. It's currently uploaded as a private test dataset here:

https://www.checklistbank.org/dataset/60320/classification

The data displays correctly in TW. I initially assumed that recent changes had corrupted their COLDP exports, but I generated the following tree directly from the export via parent relationships, so I believe the issue is in CLB.

[GLITree.txt](https://github.com/CatalogueOfLife/backend/files/12269727/GLITree.txt)

**Comment by mdoering** [comment]
When I look at the subgenus [Acontia (Acontarache) Berio, 1977](https://www.checklistbank.org/dataset/60320/taxon/267854) it claims to have an invalid parentID `267837`.
<img width="527" alt="image" src="https://github.com/CatalogueOfLife/backend/assets/327505/d99334e3-a4ec-45dc-9d4a-26f54db15ae9">

That ID does exist on Taxon.txt, but is not interpreted as it claims the nameID 23340 is invalid:
https://www.checklistbank.org/dataset/60320/verbatim/589108
<img width="464" alt="image" src="https://github.com/CatalogueOfLife/backend/assets/327505/bac1d385-e40d-437b-9ba9-0e0a4ff48a16">

This seems to be true, there is no such name record in the archive.
There are lots of records that span over several lines in the Name.txt. A TSV file does not allow that and the line breaking reference strings are likely the cause. Example from line 52426:

```
20235	20235	Ctenognophos	Prout, 1915	genus	Ctenognophos					3300	384	1915	true	ICZN	http://purl.obolibrary.org/obo/NOMEN_0000224		"233370.00|ORIGINAL GENUS UNDETERMINED|Valid name|7554.00|Geometroidea-Geometridae-Ennominae-181A-202602|Q:\Geometroidea\Geometridae\Ennominae\181A\|FC-Geometroidea-Geometridae-Ennominae-181A-202602.JPG|BC-Geometroidea-Geometridae-Ennominae-181A-202602.JPG|Fletcher (1979) stated:-
The date of publication of the part of Seitz containing page 384 in vol. 4 is stated on page 377, in both the German and English editions, to be 9.ix.1915.

The part of Seitz containing page 459 of vol. 4 (Suppl.) has appeared only in the German edition; its date of publication is stated on page 457 to be 16.x.1941. Ctenognophos was originally proposed as a subgenus of Gnophos Treitschke, 1825.
Ctenognophos was originally proposed as a subgenus of Gnophos Treitscke, 1825.

Parsons et al. (1999) included 17 species and 8 subspecies (inc. nominates) in the genus CTENOGNOPHOS.|13334|19371|Gnophos|eolaria|Guenée|1857|[INDIA]|Nord de l'Inde [northern India]|Syntypes - 2 males|(BMNH, London)|19553|by subsequent designation"	2017-07-16T21:55:16Z	Ian Kitching
20236	20236	Bizia	Walker, 1860	genus	Bizia					6336	261	1860	true	ICZN	http://purl.obolibrary.org/obo/NOMEN_0000224		233371.00|ORIGINAL GENUS UNDETERMINED|Valid name|3938.00|Geometroidea-Geometridae-Ennominae-181A-202603|Q:\Geometroidea\Geometridae\Ennominae\181A\|FC-Geometroidea-Geometridae-Ennominae-181A-202603.JPG|BC-Geometroidea-Geometridae-Ennominae-181A-202603.JPG|Parsons et al. (1999) included 2 species in the genus BIZIA.|25134|25133|Bizia|aexaria|Walker|1860|CHINA (NORTH)|Locality unknown|Holotype - male|(BMNH, London)|by monotypy	2017-07-16T21:55:16Z	Ian Kitching
```

**Comment by mdoering** [comment]
@gdower a sidenote: It would be good not to use .tsv or .txt as the file suffix when tabs are used as the delimiter. CSV should have a comma as the delimiter: https://github.com/CatalogueOfLife/coldp/blob/master/README.md#data-files

**Comment by mdoering** [comment]
Closing as this is a data formatting issue, not a backend one.

**Comment by mdoering** [comment]
@dhobern @gdower you can spot skipped records due to bad numbers of columns or empty rows through the [previous line skipped](https://www.checklistbank.org/dataset/60320/verbatim?issue=previous%20line%20skipped) issue.

**Comment by mdoering** [comment]
Note that the [RFC 4180](https://www.rfc-editor.org/rfc/rfc4180) CSV format allows for multiline content when quoted properly.

**Comment by mdoering** [comment]
https://github.com/CatalogueOfLife/coldp/blob/master/README.md#quoting-and-escaping

**Comment by dhobern** [comment]
Thanks Markus. @gdower - should I raise a TW issue?

**Comment by dhobern** [comment]
@mdoering I've just realised why I wasn't seeing the problem in my code. I use Pandas in my Python library and it handles stray line breaks on import. This means I should be able to write a really trivial processor to fix TW exports until this gets addressed there.

**Comment by gdower** [comment]
I'm also going to try to adjust the TW exporter to sanitize the line breaks (and change the file extension from csv to tsv), but it would be more user friendly if CLB can handle delimiter issues.

**Comment by gdower** [comment]
@dhobern, if you configure your datasets like this it does fix the issue as well:

![image](https://github.com/CatalogueOfLife/backend/assets/8573609/5c8da8e6-67d4-4cd2-a4b3-913fd9719d4d)

**Comment by mdoering** [comment]
> Joining the lines fixes the problem. @mdoering Many CSV libraries handle this kind of thing automatically. Should CLB do so too?
> 
> If a row includes fewer tabs/commas than the header row, try reading and joining the next line. That was all that was needed here.

I was wondering about that, probably good to have in place. On the other hand the text files are wrongly encoded and it might be good to know that. I guess we can still flag those records once we have successfully read them. I will look into this. Are you aware of a java library that does that? CLB uses the [univocity.parsers](https://github.com/uniVocity/univocity-parsers).

**Comment by dhobern** [comment]
Sorry, no - the last time I did something like this in Java, I wrote my own class that really only reflects what I needed at the time: 

https://github.com/dhobern/utils/blob/master/src/main/java/io/github/dhobern/utils/CSVReader.java

**Comment by mdoering** [comment]
Looking at TW multiline data I see it is actually properly quoted. The problem is that this is a CSV feature only and instead of the filename suggests it is a tab delimited TSV instead which should escape new lines and carriage returns with `\n` and `\r`.

**Comment by mdoering** [comment]
I have implemented support for stray away multilines as long as they end up being exactly the same number of columns as the header tells. It would not remove the TW quotes in TSV files though.

**Comment by gdower** [comment]
I did modify the TW exporter to strip `\t`, `\r\n` and `\n` in remarks, replacing them with spaces. I changed the file extension to tsv. I was also able to import @dhobern's export into dev with the csv options configured to have delimiter=\t, escape=", and enclosure=" ([as above](https://github.com/CatalogueOfLife/backend/issues/1242#issuecomment-1679279648)) but that extra dataset configuration should be unnecessary once the csv code gets deployed in TW.
